README.txt
WebHelpers
++++++++++++

WebHelpers provides functions useful in web applications: generating HTML tags,
showing results a pageful at a time, etc.  It may be used with any web
framework or template engine.  A brief summary of the module layout is below;
see the docstrings in the source for documentation.

``constants``
    Country codes, states and provinces.

``containers``
    High-level container objects and dict/list helpers.

``date``
    Date/time helpers.  These currently format strings based on dates.

``feedgenerator``
    A syndication feed library, used for generating RSS, Atom, etc.
    Ported from Django.

``html``
    A package of HTML-related helpers.

    ``html.builder``
        A library for generating HTML tags with smart escaping.  All
        public symbols are imported into ``webhelpers.html``.

    ``converters``
        Text-to-HTML converters.

    ``tags``
        High-level HTML tags, including form tags, hyperlinks, and 
        Javascript/CSS links.  The ``ModelTags`` class builds input
        tags from database records (for any kind of database).

    ``tools``
        Helpers producing chunks of HTML.

``markdown``
    A text to HTML converter.  Normally invoked via
    ``webhelpers.tools.markdown()``.  (If you use this library directly, you
    may have to wrap the results in ``literal()`` to prevent double escaping.)

``media``
    Helpers for images, PDFs, etc.

``misc``
    Miscellaneous helpers that are neither text, numeric, container, or date.

``number``
    Numeric helpers and number formatters.

``paginate``
    A tool for letting you view a large sequence a screenful at a time,
    with previous/next links.

``pylonslib``
    Helpers for the Pylons framework.  These depend on Pylons context variables
    (request, response, session, etc) or other aspects of Pylons and its
    dependencies.  Most can be ported to other frameworks with little effort.
    
``string24``
    The ``string`` module from Python 2.4.  Useful if you're running on
    Python 2.3.

``tags``
    Helpers producing simple HTML tags.

``text``
    Helpers producing string output, suitable for both HTML and non-HTML
    applications.

``textile``
    Another text to HTML converter.  Normally invoked via
    ``webhelpers.tools.textilize()``.  (If you use this library directly, you
    may have to wrap the results in ``literal()`` to prevent double escaping.)

``util``
    Miscellaneous functions.

WebHelpers is package aimed at providing helper functions for use within web
applications.

These functions are intended to ease web development with template languages by
removing common view logic and encapsulating it in re-usable modules as well as
occasionally providing objects for use within controllers to assist with common
web development paradigms.

For support/question/patches, please use the `Pylons mailing list
<http://groups.google.com/group/pylons-discuss>`_.

Requirements
------------

WebHelpers does not have any install dependencies, but some functions depend
on third-party libraries (Routes_, Pylons_, WebOb_, unidecode_) as specified in
their documentation.


.. _Routes:  http://routes.groovie.org/
.. _Pylons:  http://pylonshq.com/
.. _WebOb:   http://pythonpaste.org/WebOb
.. _unidecode:  http://python.org/pypi/Unidecode/


CHANGELOG
tip (XXX-XX-XX)
-----------------

* webhelpers.html.converters: bugfix in ``format_paragraphs(), should return a
  literal. (#74)

1.3 (2011-03-24)
------------------

* webhelpers.paginate:

  - Revert Sprox patch (#59) from 1.3b1; it makes SQLAlchemy inefficient (#63).
    The bug reporter's Sprox object seems to be unsliceable and therefore
    incompatible with paginate. Changed error message to say this if the
    slicing operation raises "TypeError: unhashable type".

1.3b1 (2011-03-17)
------------------

* Documentation:
  
  - Some ``webhelpers.misc`` helpers were undocumented.
  
  - Spelling corrections throughout, done by Marius Gedminas.

* webhelpers.date:

  - Adjust test in 'test_date.py' to account for leap years. (#61, reported
    by Andrey Rahmatullin / wrar)

* webhelpers.html.grid, webhelpers.pylonslib.grid:

   - Add 'request' and 'url' args to support paged display.
     (webhelpers.html.grid only)
   - Deprecate webhelpers.pylonslib.grid. It may be removed in a future
     version.
  
* webhelpers.paginate:

  - Add URL generator classes for new frameworks like Pyramid.
  - In the Pylons backward compatibility code, support page parameter in the
    route path (e.g., "/help/page5") when using Routes explicit mode without a
    URL generator.

* setup.py:

   - Fix typo in test requirements. (#60, Andrey Rahmatullin/wrar)

1.2 (2010-08-18)
----------------

* WebHelpers now depends on MarkupSafe. ``literal`` and ``escape`` now use it.
  
* webhelpers.html.builder:

  - ``literal`` and ``escape`` now use MarkupSafe, which has a C speedup for
    escaping, escapes single-quotes for security, and adds new methods to
    ``literal``. Compatibility should not be a problem; but see the docs if
    you encounter any edge cases.

* webhelpers.html.tags:

  - For new HTML 5 input fields, the ``text`` helper has a "type" argument.

  - Don't put an "id" attribute on a hidden fields generated by the ``form()``
    helper, including the magic ``_method`` field. The IDs will clash if there
    are multiple forms on the page.

* webhelpers.html.tools:

  - Preserve case of "method" arg in ``button_to()`` for XHTML compatibility. 
    Patch by transducer.

* webhelpers.text:

  - Urlencode ``urlify`` return value in case it contains special
    characters like "?". Reported by mcd34@gmail.com.

* webhelpers.util:

  - Fix bug in ``update_params`` in handling existing query strings. Support
    multiple values per parameter.

1.1 (2010-08-09)
----------------

* webhelpers.paginate:

  - Remove stray Routes import. (Other conditional Routes imports remain for
    backward compatibility; see module docstring.)

* webhelpers.pylonslib.minify:

  - Remove _jsmin module due to licensing issues. Details are in
    webhelpers/pylonslib/_jsmin.py .  You can install the "jsmin" package in
    PyPI (which has the same license), and the helper will use it. If that
    package is installed, the helper will use it.  Otherwise the helper will
    emit a warning and leave the Javascript unchanged. CSS minification is not
    affected.


1.0 (2010-06-01)
-----------------

* webhelpers.html.tools:

 - Bugfix re URLs surrounded by [].  Bug #32.

1.0rc1 (2010-05-24)
-------------------

* webhelpers.html.tags:

 - Change 'id' argument to all form field helpers. The default value is now
   NotGiven rather than None. NotGiven tells WebHelpers to create an ID
   attribute based on the field name (formerly None did this). None suppresses
   the ID attribute entirely (formerly "" did this, and still does for backward
   compatibility).  This behavior of None is consistent with other parts of
   WebHelpers.

* webhelpers.misc:

 - New ``format_exception`` helper to display an exception as Python would but
   without the traceback.

1.0b7 (2010-05-16)
------------------

* webhelpers.containers:

  - Bugfix in ``canada_provinces``, reported by rpetrello.

* webhelpers.html.grid / webhelpers.pylonslib.grid:

  - Updates by Ergo, mainly styling and CSS classes.
  - Rename classes: ObjectGrid, PylonsGrid, PylonsObjectGrid.

* webhelpers.paginate:

 - New URL generation algorithm for ``Page.pager()``. You can pass a callback
   function to the constructor, or it will fall back to ``pylons.url.current``
   or ``routes.url_for`` (in that order). It will raise ``NotImplementedError`` if
   none of these are available.
 - Don't allow extra positional args in constructor. The implementation does
   nothing with them, so it shouldn't allow them.
 - Import ``sqlalchemy.orm`` as well as ``sqlalchemy``. User Sybiam reports an
   error otherwise.
 - Add code to work with other iterable containers, contributed by Marcin
   Kuzminski.

* webhelpers.pylonslib.flash:

  - New argument ``ignore_duplicate`` to prevent adding the same message
    multiple times.


1.0b6 (2010-04-23)
------------------

* webhelpers.containers / webhelpers.misc:

  - ``NotGiven`` moved to webhelpers.misc.

* webhelpers.html.grid / webhelpers.pylonslib.grid:

  - Updates by Ergo, including SQLAlchemy object grid classes.

* webhelpers.misc:

  - New function ``deprecate``.

* webhelpers.number:

  - New functions ``format_data_size``, ``format_byte_size``, and
    ``format_bit_size`` for displaying numbers in SI units
    ("1.2 kilobytes", "1.2 kB", "1.0 KiB"). Contributed by
    Wojciech Malinowski.

1.0b5 (2010-03-18)
------------------

* webhelpers.html.converters:

  - Re-add import of ``render`` and ``sanitize`` from
    ``webhelpers.html.render``.  That module is not public.

* webhelpers.misc:

  - New exception ``OverwriteError``.
  - Add ``exclude`` argument to ``subclasses_only``.

* webhelpers.text:

  - Disable ``convert_misc_characters``: it fails its doctests and there's
    no consensus on what it should do.

* "number_to_human_size.py" is in unfinished directory. This is an old rails
  helper from WebHelpers 0.6.4. It's here pending a more comprehensive helper;

webhelpers/urls.py
"""Url helpers"""
import urllib
import re

from webhelpers.asset_tag import compute_public_path
from webhelpers.html import HTML, escape, literal

__all__ = ['link_to', 'button_to', 'mail_to']


def link_to(name, url='', **html_options):
    """
    Create link tag with text ``name`` and a URL created by the set of
    ``options``.
    
    See the valid options in the documentation for Routes url_for.
        
    Optionally you can make the link do a POST request (instead of the
    regular GET) through a dynamically added form element that is
    instantly submitted. Note that if the user has turned off
    Javascript, the request will fall back on the GET. So its your
    responsibility to determine what the action should be once it
    arrives at the controller.
    
    The POST form is turned on by passing ``post`` as True. Note, it's
    not possible to use POST requests and popup targets at the same
    time (an exception will be thrown).
    
    Examples::
    
        >> link_to("Delete this page", url(action="destroy", id=4))
        >> link_to("Destroy account", url(action="destroy"), 
        .. method='delete')
        
    """
    if callable(url):
        url = url()
    html_options['href'] = url
    return HTML.a(name or url, **html_options)


def convert_boolean_attributes(html_options, bool_attrs):
    """Utility function to convert boolean attributes into proper
    HTML attribute
    
    For example, this will convert ``selected=True`` into
    ``selected="selected"``.
    
    """
    for attr in bool_attrs:
        if html_options.has_key(attr) and html_options[attr]:
            html_options[attr] = attr
        elif html_options.has_key(attr):
            del html_options[attr]


def button_to(name, url='', **html_options):
    """Generate a form containing a sole button that submits to
    ``url``. 
    
    Use this method instead of ``link_to`` for actions that do not have
    the safe HTTP GET semantics implied by using a hypertext link.
    
    The parameters are the same as for ``link_to``.  Any 
    ``html_options`` that you pass will be applied to the inner
    ``input`` element. In particular, pass
    
        disabled = True/False
    
    as part of ``html_options`` to control whether the button is
    disabled.  The generated form element is given the class
    'button-to', to which you can attach CSS styles for display
    purposes.
    
    The submit button itself will be displayed as an image if you 
    provide both ``type`` and ``src`` as followed:

         type='image', src='icon_delete.gif'

    The ``src`` path will be computed as the image_tag() computes its 
    ``source`` argument.

    Example 1::
    
        # inside of controller for "feeds"
        >> button_to("Edit", url(action='edit', id=3))
        <form method="POST" action="/feeds/edit/3" class="button-to">
        <div><input value="Edit" type="submit" /></div>
        </form>
    
    Example 2::
    
        >> button_to("Destroy", url(action='destroy', id=3), 
        .. method='DELETE')
        <form method="POST" action="/feeds/destroy/3" 
         class="button-to">
        <div>
            <input type="hidden" name="_method" value="DELETE" />
            <input value="Destroy" type="submit" />
        </div>
        </form>

    Example 3::

        # Button as an image.
        >> button_to("Edit", url(action='edit', id=3), type='image', 
        .. src='icon_delete.gif')
        <form method="POST" action="/feeds/edit/3" class="button-to">
        <div><input alt="Edit" src="/images/icon_delete.gif"
         type="image" value="Edit" /></div>
        </form>
    
    .. note::
        This method generates HTML code that represents a form. Forms
        are "block" content, which means that you should not try to
        insert them into your HTML where only inline content is
        expected. For example, you can legally insert a form inside of
        a ``div`` or ``td`` element or in between ``p`` elements, but
        not in the middle of a run of text, nor can you place a form
        within another form.
        (Bottom line: Always validate your HTML before going public.)
    
    """
    if html_options:
        convert_boolean_attributes(html_options, ['disabled'])
    
    method_tag = ''
    method = html_options.pop('method', '')
    if method.upper() in ['PUT', 'DELETE']:
        method_tag = HTML.input(
            type='hidden', id='_method', name_='_method', value=method)
    
    form_method = (method.upper() == 'GET' and method) or 'POST'
    
    url, name = url, name or url
    
    submit_type = html_options.get('type')
    img_source = html_options.get('src')
    if submit_type == 'image' and img_source:
        html_options.update(dict(type=submit_type, value=name,
                                 alt=html_options.get('alt', name)))
        html_options['src'] = compute_public_path(img_source, 'images', 'png')
    else:
        html_options.update(dict(type='submit', value=name))
    
    return HTML.form(method=form_method, action=url, class_="button-to",
                     c=[HTML.div(method_tag, HTML.input(**html_options))])


def mail_to(email_address, name=None, cc=None, bcc=None, subject=None, 
    body=None, replace_at=None, replace_dot=None, encode=None, **html_options):
    """Create a link tag for starting an email to the specified 
    ``email_address``.
    
    This ``email_address`` is also used as the name of the link unless
    ``name`` is specified. Additional HTML options, such as class or
    id, can be passed in the ``html_options`` hash.
    
    You can also make it difficult for spiders to harvest email address
    by obfuscating them.
    
    Examples::
    
        >>> mail_to("me@domain.com", "My email", encode = "javascript")
        literal(u'<script type="text/javascript">\\n//<![CDATA[\\neval(unescape(\\'%64%6f%63%75%6d%65%6e%74%2e%77%72%69%74%65%28%27%3c%61%20%68%72%65%66%3d%22%6d%61%69%6c%74%6f%3a%6d%65%40%64%6f%6d%61%69%6e%2e%63%6f%6d%22%3e%4d%79%20%65%6d%61%69%6c%3c%2f%61%3e%27%29%3b\\'))\\n//]]>\\n</script>')
    
        >>> mail_to("me@domain.com", "My email", encode = "hex")
        literal(u'<a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;%6d%65@%64%6f%6d%61%69%6e.%63%6f%6d">My email</a>')
    
    You can also specify the cc address, bcc address, subject, and body
    parts of the message header to create a complex e-mail using the 
    corresponding ``cc``, ``bcc``, ``subject``, and ``body`` keyword 
    arguments. Each of these options are URI escaped and then appended
    to the ``email_address`` before being output. **Be aware that 
    javascript keywords will not be escaped and may break this feature 
    when encoding with javascript.**
    
    Examples::
    
        >>> mail_to("me@domain.com", "My email", cc="ccaddress@domain.com", bcc="bccaddress@domain.com", subject="This is an example email", body= "This is the body of the message.")
        literal(u'<a href="mailto:me@domain.com?cc=ccaddress%40domain.com&amp;body=This%20is%20the%20body%20of%20the%20message.&amp;subject=This%20is%20an%20example%20email&amp;bcc=bccaddress%40domain.com">My email</a>')
        
    """
    extras = {}
    for key, option in ('cc', cc), ('bcc', bcc), ('subject', subject), \
                       ('body', body):
        if option:
            if not isinstance(option, literal):
                option = escape(option)
            extras[key] = option
    options_query = urllib.urlencode(extras).replace("+", "%20")
    protocol = 'mailto:'

    email_address_obfuscated = email_address
    if replace_at:
        email_address_obfuscated = email_address_obfuscated.replace('@', 
            replace_at)
    if replace_dot:
        email_address_obfuscated = email_address_obfuscated.replace('.', 
            replace_dot)

    if encode == 'hex':
        email_address_obfuscated = HTML.literal(''.join(
            ['&#%d;' % ord(x) for x in email_address_obfuscated]))
        protocol = HTML.literal(''.join(['&#%d;' % ord(x) for x in protocol]))

        word_re = re.compile('\w')
        encoded_parts = []
        for x in email_address:
            if word_re.match(x):
                encoded_parts.append('%%%x' % ord(x))
            else:
                encoded_parts.append(x)
        email_address = HTML.literal(''.join(encoded_parts))

    url = HTML.literal(protocol + email_address)
    if options_query:
        url += HTML.literal('?') + options_query
    html_options['href'] = url

    tag = HTML.a(name or email_address_obfuscated, **html_options)

    if encode == 'javascript':
        tmp = "document.write('%s');" % tag
        string = ''.join(['%%%x' % ord(x) for x in tmp])
        return HTML.script(
            HTML.literal("\n//<![CDATA[\neval(unescape('%s'))\n//]]>\n" % string),
                         type="text/javascript")
    else:
        return tag
